src/main/java/org/truffleruby/language/loader/FeatureLoader.java

```diff
@@ -397,9 +407,11 @@ private String findFeatureWithAndWithoutExtension(String path) {
             return asCExt;
         }
 
-        final String withoutExtension = findFeatureWithExactPath(path);
-        if (withoutExtension != null) {
-            return withoutExtension;
+        if (!triedWithoutExtension) {
+            final String withoutExtension = findFeatureWithExactPath(path);
+            if (withoutExtension != null) {
+                return withoutExtension;
+            }
         }
```

Comment:
I wonder if this case is actually needed semantically, i.e. can `require` ever load anything which is not a `.rb` or a `.so/bundle/dll`?
It seems not:
```
$ echo 'p :a' > a
$ ruby -I. -e 'require "a"'
cannot load such file -- a (LoadError)

$ mv a a.rb
$ ruby -I. -e 'require "a"'
:a
```

There are specs for this kind of stuff in spec/ruby/core/kernel/shared/require.rb.
It would be good to add one for this case if there is not already one.

Given we already has this bug before, and there is no tag in spec/tags/core/kernel/require_tags.txt that sounds like it it seems there is no spec for this already, so could you add one, and then remove this part?

Reply:
Ah but my example above uses $LOAD_PATH and `findFeatureWithAndWithoutExtension()` is not used for $LOAD_PATH search.
However the behavior is as I said above:
```
$ ruby -v --disable-gems -I. -e 'require "#{Dir.pwd}/a"'
ruby 3.2.2 (2023-03-30 revision e51014f9c0) [x86_64-linux]
-e:1:in `require': cannot load such file -- /home/eregon/tmp/loaddir/a (LoadError)

$ ruby -v --disable-gems -I. -e 'require "#{Dir.pwd}/a"'
truffleruby 24.0.0-dev-54e96355, like ruby 3.2.2, GraalVM CE Native [x86_64-linux]
:a
```
i.e. there we are not compatible with CRuby, and this shows we don't need to try the path as-is if it doesn't end in `.rb`, `.so` or `.bundle`.
But we still need to check if it ends in `.bundle`.
So how about merging the first 2 `if`s?

Reply:
BTW I also noticed that files like `a.rb.rb` don't get loaded on CRuby by `require "#{Dir.pwd}/a.rb"`, only by `require "#{Dir.pwd}/a.rb.rb"`. So we could also structure this like:
```
if hasExtension(path)
  try as-is, and on macOS replace .so by .bundle (translateIfNativePath does that)
else
  try + .rb
  try + .CEXT_EXTENSION
end
```

Reply:
Thanks, that makes sense to me!

I was curious if there was a way to test that the java code wasn't trying extra calls (like "a.rb.rb").

We have a spec for absolute paths with .rb
I added one for absolute path with .so on darwin
and 2 for absolute paths without extensions (rb and C).

Are we missing anything?

Reply:
It looks like it doesn't like trying to load C-named files that are just text files:

```
Kernel.require (path resolution) loads c-extension file when passed absolute path without extension when no .rb is present ERROR
RuntimeError: /rws-data/repos/truffleruby-ws/truffleruby/spec/ruby/fixtures/code/a/load_fixture.so: file too short
```

Reply:
It looks like CRuby doesn't pass those specs: https://github.com/oracle/truffleruby/actions/runs/6364893890/job/17281491472?pr=3276, maybe there is different behavior for require and load :/

Probably we should move the specs which differ behavior for both to :kernel_require instead of :kernel_require_basic.

Reply:
Given both CRuby and TruffleRuby error with a /file too short/ exception in that case it seems reasonable to check that.

Reply:
> I was curious if there was a way to test that the java code wasn't trying extra calls (like "a.rb.rb").

I'm not aware of one, that seems difficult to test. We could add a benchmark perharps but that would likely be noisy based on disk access, etc.

Reply:
I moved the specs around and fixed them and did some trivial optimization, now all should be green :)

Reply:
Great, thanks for your help!